### ISSUE TITLE
integration into quanteda as a core tokenizer

### ISSUE THREAD

**Issue by kbenoit** [ropensci/tokenizers#25]
I started to continue our comments on #24 but thought it best to start a new issue.

As for **quanteda**, we are thinking of an rOpenSci-type overhaul of the API that would be a major change. (The cleanness of the tokenizers API is an inspiration.) See https://github.com/kbenoit/quanteda/blob/quanteda0_9_9/API.md. Comments most welcome.

There are some API changes I might suggest to **tokenizers**, but minor ones. The main thing is figuring out whether we need to wrap the **tokenizers** in order to class them for **quanteda** objects, or to extend them with options in some way that get classed so that they can integrate better.

There are other issues too such as the exceptions needed for tokening Twitter text, which standard (**stringi**) methods tend to obliterate: `#` and `@` in particular.

**Comment by lmullen** [comment]
@kbenoit I like where you are going with your API. It seems like wrapping tokenizers would best happen inside quanteda. But if you think it would be helpful to provide conversion functions or the like directly in tokenizers, please let me know.

Is there already a hashtag/twitter tokenizer in quanteda? We could add one to tokenizers if you like.

**Comment by kbenoit** [comment]
OK, will do on the possible **tokenizers** tweaks that would aid integration. What's the best way? a list added to this issue?

`quanteda::tokenize()` calls many of the same **stringi** functions that your package does, but with more options. We preserve the special Tweet characters through substitution prior to punctuation-obliterating tokenizing, and then replace. Slow but I am not sure what the alternatives are.

**Comment by lmullen** [comment]
Sure, a list added to this issue would be fine.

The tokenizer for Twitter should be its own function: `tokenize_tweets()` or something like that. 

Instead of substituting, have you tried stripping only some of the punctuation so that you leave `@#`, then splitting on white space? Might be faster and yet reliable enough.

**Comment by lmullen** [comment]
@kbenoit Gearing up for another release of tokenizers and thought it would be good to take on this issue. What do you think would make tokenizers more useful for quanteda? Happy to wait on this till after the conference if you think we will take this up there.

**Comment by kbenoit** [comment]
OK, we've been thinking about it and discussing it and here is the best current summary of my thoughts. It's a lot, and I would be happy to create separate issues as you prefer, but let me know which are worth pursuing.

## Ways to implement segmentation

The basic point of a tokenizer is to segment a text into word, character, or sentence units. (We could add others to that but that's the basic unit target set based on linguistic definitions.) We would love to use **tokenizers** for this in **quanteda**, which is entirely possible since our `tokens()` function has a completely detachable segmentation step, before it implements special handling of tokens objects by hashing them, something we do to make subsequent processing operations more efficient.

### different methods

Here are some simple segmentation methods:

*  on " " - this is the the (literally, not pejoratively) stupidest way, but it's fast. This is the `quanteda::tokens(x, what = "fastestword")` and calls `stringi::stri_split_fixed(x, " ")`.

*  on a whitespace character class - just about as fast above is `quanteda::tokens(x, what = "fasterword")`, which calls `stringi::stri_split_charclass(txt, "\\p{WHITE_SPACE}")`. (This could be done using `tokenizers::tokenize_regex(x, pattern = "\\p{WHITE_SPACE}")`).

*  using the much smarter boundaries definitions, which is used in `tokenizers::tokenize_words(x)` and in `quanteda::tokens(x, what = "word")`, both calling `stringi::stri_split_boundaries(txt, type = "word")`. 

### segmentation versus transformation and selection

There are options in `stringi::stri_split_boundaries(..., type = "word")` that remove punctuation  and symbols (`skip_word_none`), and `skip_word_number` to control how numerals/numbers are handled. These combine segmentation with selection, but are so fast that it's worth combining them. However this should be done through some more consistent interface such as `skip_something` or `remove_punct`. 

My own view is that changing tokens (the case for instance) is a transformation separate from tokenizing. Maybe selecting elements from the tokens, or ignoring some tokens is also separate, although it also goes to the definition of tokens. (and as I note above these are integrated into `stringi::stri_split_boundaries()` and therefore really efficient. Removing stopwords also has nothing to do with tokenization, and my suggestion would be to make this a separate function. I'm not hung up on any of this for **tokenizers** however since I just want the segmentation functions to work to feed into **quanteda**, and can keep token modification and selection separate (as they are now in **quanteda**).

### summing up

Our existing code for tokenization works in **quanteda** but it's a mess and difficult to make modular. We'd love to make the **tokenizers** functions the default segmentation functions, but also have the option to use others (**spacyr**, for instance, or just the dumb whitespace splitters if those are enough and speed is really important.)

To make a long story short, I'd like to separate segmentation from the `quanteda::tokens()` constructor and this could very easily use the `tokenisers::tokenize_*` functions. We want to, but need a bit of additional functionality first. I've listed these below, but would be happy to split these into separate issues.


## Wish list

1. Would like an option to preserve punctuation.

    ```{r}
    txt <- "Hey: Y, M, C, A!!"
    # currently
    tokenize_words(txt, lowercase = FALSE)
    # want this
    tokens(txt, remove_punct = TRUE)
    tokens(txt, remove_punct = FALSE)
    ``` 

    This might also separate the definitions of punctuation from "symbols", since "$100" is different from "100" and we might want to remove punctuation without removing the "$".

    It looks like `tokenizers::tokenize_ptb()` can do a lot of this, but it looks like it would take a strong performance hit. I have not run (m)any tests on this because it's such a recent addition to **tokenizers**.

2. Need an option for preserving Twitter characters, e.g.

    ```{r}
    # currently
    tokenize_words("@kenbenoit loves #rstats!")
    # want this
    tokens("@kenbenoit loves #rstats!", remove_punct = TRUE, remove_twitter = FALSE)
    ```
    We do this through a somewhat expensive process of substitute, tokenize, replace.  It might be faster to segment without removing punctuation first, and then remove puctuation except words starting with `#` or `@`.
 
3. We'd like to be able to preserve intra-word hyphens, e.g.

    ```r
    tokenize_words("Keep co-operate as one word.", split_hyphenated = FALSE)
    [[1]]
    [1] "keep"    "co-operate" "as"      "one"     "word"
    ```
    
4. We would like an option to keep intra-token separators.

    ```r
    # currently
    tokenize_words("one\ttwo\n  three four")
    # want this
    tokens("one\ttwo\n  three four", remove_separators = TRUE)
    tokens("one\ttwo\n  three four", remove_separators = FALSE)
    ``` 
    Why?  If move to a structure where a corpus is an indexed set of tokens, then to reconstruct the texts we need the original spaces, including knowing what was a `" "`, versus `"\n"`, etc.

5. `tokenize_sentences()`:  Add an exception list, similar to stopwords, that could be used for exceptions to sentence segmentation.  For instance:

    ```r
        # Replace . delimiter from common title abbreviations, with _pd_
        exceptions <- c("Mr", "Mrs", "Ms", "Dr", "Jr", "Prof", "Ph.D", "M", "MM", "St", "etc")
        findregex <- paste0("\\b(", exceptions, ")\\.")
        txt <- stri_replace_all_regex(txt, findregex, "$1_pd_", vectorize_all = FALSE)
        
        ## Remove newline chars 
        txt <- lapply(txt, stringi::stri_replace_all_fixed, "\n", " ")
        
        ## Perform the tokenization
        tok <- stringi::stri_split_boundaries(txt, type = "sentence")
        
        ## Cleaning
        tok <- lapply(tok, function(x){
            x <- x[which(x != "")] # remove any "sentences" that were completely blanked out
            x <- stringi::stri_trim_right(x) # trim trailing spaces
            x <- stri_replace_all_fixed(x, "_pd_", ".") # replace the non-full-stop "." characters
            return(x)
        } )
    ```
    
6. Some URL handling would be nice

    ```r
    txt <- "The URL is http://textworkshop17.ropensci.org#schedule."
    # want this
    tokenize_words(txt, lowercase = FALSE, strip_url = TRUE)
    [[1]]
    [1] "The"            "URL"            "is"
    tokenize_words(txt, lowercase = FALSE, strip_url = FALSE)
    [[1]]
    [1] "The"            "URL"            "is"
    [4] "http://textworkshop17.ropensci.org#schedule"
    ``` 

    Tweets, for instance, are replete with URLs.


We've explored these functionality differences at https://github.com/kbenoit/quanteda/blob/master/tests/misc/tokens_comparison_to_tokenizers.md which includes some benchmarks too.

## A note on nomenclature

As **tokenizers** expands, I'd consider keeping a tight control on the logic of the function names. With some recent additions, the function names are split between using the second part of the function name as the unit of tokenisation, and using it to describe the segmentation method.

`tokenize_words()` tokenises a text into words, but `tokenize_regex()` tokenises into (in the name) unspecified units using a regular expression to identify the delimiters.

**Comment by kbenoit** [comment]
BTW I like the idea of a separate segmenter for Tweets. My inner autistic child is cringeing though if it's called `tokenize_tweets` however since that refers to a third element, which is the text type. 

But the idea is good since it could use special handling for @names #hashtags and URLs, which would be unnecessary and costly in other contexts.

**Comment by lmullen** [comment]
@kbenoit Thanks for thinking this through in such detail. I've been mulling this over today, and here are some thoughts.

First, I think that all of these suggestions will have to go into tokenizers v0.3. We're close to a v0.2 release with the corrected skip n-grams and other improvements, and I'd like to get that out the door. But this will give us time to work out how to implement you suggestions.

Second, I agree in principle with all the desires for features, but I have some slightly different ideas about the user interface. I'm trying to think about this in a way that prioritizes performance and usability, while minimizing breaking changes.

For now, let me focus on the user interface question before we get to the details of how to, e.g., do the tokenization for tweets. I agree that segmentation is conceptually separate from tokenization, and also that normalization (lowercasing, removing punctuation) is conceptually separate from tokenization and segmentation. However, there are potentially significant performance implications for separating out all of these, since if you can loop over the text one time rather than two or three, you can save time.

Another consideration: Some packages tend towards having lots of arguments to a function. Quanteda's `tokens()` function tends in that direction at the moment. That approach can be very usable, especially for high level packages. Other packages tend to have lots of different functions. I think stringi is an example of a package which is far less usable for separating so many things into different functions. But I've come to appreciate the usefulness of how @dselivanov has designed text2vec with composable functions, where it is possible to substitute one's own functions rather than the one's in the package. 

So here is a proposal for the v0.3 interface that takes those considerations into account.

## Proposal

1. Segmentation will be separated from tokenization. New functions will be added that do the segmentation. An incomplete list (names up for negotiation).

- `segment_word_boundaries()`: wrapper around stringi's `stri_split_boundaries(type = "word")
- `segment_space()`:  quanteda's `fastestword`
- `segment_whitespace()`: quanteda's `fasterword`
- `segment_for_tweets()`: This will be a custom segmenter which takes into account hashtags, @usernames, URLs, etc. This will have to be written in Rcpp for performance.
- `segment_chars()`
- `segment_ptb()`: Penn Treebank seems more like a segmenter in this way of thinking about things than a tokenizer.

2. Tokenization will be the process of assembling segments into tokens. Segmentation functions will be passed into tokenization functions. So the (partial) signature for a function will look like this.

```
tokenize_ngrams <- function(x, segment = tokenizers::segment_word_boundaries, n = 3) {
  # check x for validity
  segments <- segment(x)
  out <- tokenize_ngrams_vectorized(segments) # the Rcpp function
  out
}
```

This is important because from a user's perspective they only need to keep using the same functions, but now they have the flexibility to pass in whatever segmenter they want.

3. Stop words will continue to be handled at the tokenization phase. This is simply because they can be done much more efficiently when you are already looping over the segments.

4. Normalization will likely happen at one of those stages. I haven't decided which: depends on where it is more efficient. But there is nothing to prevent someone from, e.g., lowercasing the tokens themselves at a later stage, or stripping out punctuation. They are just character vectors. And it's possible that I can make normalization efficient.

5. Anyone who wants a simple function rather than the composability of those functions can easily write a wrapper. We might include specialized functions with sane defaults for certain tasks, like tweets.

What do you think? I think that design will let us accomplish everything in your wish list with a reasonable interface and keeping good performance.

**Comment by Ironholds** [comment]
(Chipping in pseudorandomly): this makes sense to me, although tweet-handling seems a very specific use case. My one pedantic comment is I think the segmenter params should go at the end - probabilistically it's the default value the base user is least likely to want to mess with, so moving it to the end allows for nameless values for other components.

**Comment by kbenoit** [comment]
Great discussion started here and look forward to discussing further in 2 weeks!

**Comment by kbenoit** [comment]
Great discussion here.

A few thoughts on the part before the **Proposal** section:

-  (As you know) I wrote a `tokenize_tweets()` that is consistent with the API for v0.2 but inconsistent with below, but we can modify this once the v0.3 API issues are decided (as one of several functions that will change).

-   On the complexity of `quanteda::tokens()`:  We agree, it’s become too complex.  Our next iteration will remove nearly all of the `remove_whatever` options and add a new argument called `tokeniser = <function>` which along with `…` will pass through options specific to a external tokenizer function.  We’d like most of those to come from **tokenizers** but also plan to have wrappers to say `spacyr::spacy_parse()`.

    For this new usage, I’m actually happy with functions such as `tokenize_whitespace()` or `tokenize_tweets()`.  But we’re still thinking about the best mixture of functions that tokenise in a specific manner, and functions based on a specific technology that have a similar mix of options (e.g. `remove_punct` or `strip_punctuation`).

    I must admit however that I’m a huge fan of the **stringi** function name structure: I find it exceptionally clear to understand the functions and how they are different, and also to remember them.  When we overhauled the **quanteda** function names last year, we had a similar “consistent grammar” in mind.  I also wanted the package help index to nicely organise the hierarchy of functions, hence the anal-retentive (to some!) use of `data_corpus_inaugural` etc.

On the proposal:

-   I like the idea, since this provides the building blocks for others to assemble in slightly different ways if needed.  The key question I think is whether or how to handle steps such as options to keep whitespace, punctuation characters, or numbers and numerals.  I’d be happy to write options for these into `tokenize_words()` but only after you’ve approved an API design.  This choice is entirely yours.  I could always write a wrapper in **quanteda** around the more basic functions in **tokenizers**, but then I’m losing the some advantage of being able to call the **tokenizers** functions rather than have **quanteda**’s own set of tokenizers.

-   I’m not sure I understand point 3 about it being more efficient to remove stopwords when there is already a loop.

    From `basic-tokenisers.R` and `tokenize_words()`:

    ```r
      if (!is.null(stopwords)) out <- lapply(out, remove_stopwords, stopwords)
    ```

    is after the `stri_split_boundaries()` call has returned a list, so there is no efficiency gain to adding this `lapply()` call to the `tokenize_words()` function versus making it separate.

    `tokenize_word_stems()` is the same, it’s a call to `tokenize_words()` and then an `lapply()` to `SnowballC::wordStem()`.  No efficiency gain, and means that you have a second function to add a third step to a first function that performs steps 1 and 2 (tokenizing and removing stop words).

    Thinking about the logic of this, it would be an equally valid approach to have a `tokenize_words(x, wordstem = TRUE)` (or some variation) with a second function wrapping to this called `tokenize_stopwords()`. Because both stopword removal and word stemming are separate `lapply()` operations, there is indeterminacy as to which belongs as a separate “tokenizer” and which belongs as an option to the word tokeniser.  (My preferred solution of course is to have both be separate from the tokeniser function.)

-   Echoing my points above and made elsewhere, I am more concerned about how to handle the non-word characters such as punctuation, numerics, and whitespace than normalising word features.  Those definitely can be more efficient as tokeniser options and are uncontroversial core tokenisation functions.



On 6 Apr 2017, at 18:03, Lincoln Mullen <notifications@github.com<mailto:notifications@github.com>> wrote:


@kbenoit<https://github.com/kbenoit> Thanks for thinking this through in such detail. I've been mulling this over today, and here are some thoughts.

First, I think that all of these suggestions will have to go into tokenizers v0.3. We're close to a v0.2 release with the corrected skip n-grams and other improvements, and I'd like to get that out the door. But this will give us time to work out how to implement you suggestions.

Second, I agree in principle with all the desires for features, but I have some slightly different ideas about the user interface. I'm trying to think about this in a way that prioritizes performance and usability, while minimizing breaking changes.

For now, let me focus on the user interface question before we get to the details of how to, e.g., do the tokenization for tweets. I agree that segmentation is conceptually separate from tokenization, and also that normalization (lowercasing, removing punctuation) is conceptually separate from tokenization and segmentation. However, there are potentially significant performance implications for separating out all of these, since if you can loop over the text one time rather than two or three, you can save time.

Another consideration: Some packages tend towards having lots of arguments to a function. Quanteda's tokens() function tends in that direction at the moment. That approach can be very usable, especially for high level packages. Other packages tend to have lots of different functions. I think stringi is an example of a package which is far less usable for separating so many things into different functions. But I've come to appreciate the usefulness of how @dselivanov<https://github.com/dselivanov> has designed text2vec with composable functions, where it is possible to substitute one's own functions rather than the one's in the package.

So here is a proposal for the v0.3 interface that takes those considerations into account.

Proposal

  1.  Segmentation will be separated from tokenization. New functions will be added that do the segmentation. An incomplete list (names up for negotiation).

  *   segment_word_boundaries(): wrapper around stringi's `stri_split_boundaries(type = "word")
  *   segment_space(): quanteda's fastestword
  *   segment_whitespace(): quanteda's fasterword
  *   segment_for_tweets(): This will be a custom segmenter which takes into account hashtags, @usernames<https://github.com/usernames>, URLs, etc. This will have to be written in Rcpp for performance.
  *   segment_chars()
  *   segment_ptb(): Penn Treebank seems more like a segmenter in this way of thinking about things than a tokenizer.

  1.  Tokenization will be the process of assembling segments into tokens. Segmentation functions will be passed into tokenization functions. So the (partial) signature for a function will look like this.

tokenize_ngrams <- function(x, segment = tokenizers::segment_word_boundaries, n = 3) {
  # check x for validity
  segments <- segment(x)
  out <- tokenize_ngrams_vectorized(segments) # the Rcpp function
  out
}

This is important because from a user's perspective they only need to keep using the same functions, but now they have the flexibility to pass in whatever segmenter they want.

3. Stop words will continue to be handled at the tokenization phase. This is simply because they can be done much more efficiently when you are already looping over the segments.

4. Normalization will likely happen at one of those stages. I haven't decided which: depends on where it is more efficient. But there is nothing to prevent someone from, e.g., lowercasing the tokens themselves at a later stage, or stripping out punctuation. They are just character vectors. And it's possible that I can make normalization efficient.

5. Anyone who wants a simple function rather than the composability of those functions can easily write a wrapper. We might include specialized functions with sane defaults for certain tasks, like tweets.

What do you think? I think that design will let us accomplish everything in your wish list with a reasonable interface and keeping good performance.

**Comment by lmullen** [comment]
@kbenoit You're right to point out that those methods of removing the stopwords are inefficient. Those are some potentially slow parts of the code that I would like to evaluate. In my comment about efficiency, I was thinking a better model was what is being done in skip-n-grams, e.g., https://github.com/ropensci/tokenizers/blob/master/src/skip_ngrams.cpp#L11 But that's all a very minor point, and I intend to benchmark it eventually to see whether my hunches about performance are correct or not.

I think the much more important point, which most of the discussion above tends towards, is getting the interface right. If we can get the interface right, then the implementation details can follow.

I think I'm leaning towards your comment about about the usefulness of the stringi approach, having multiple functions. It seems like what I was proposing above in passing a function to segment then to tokenize might not work very well, since in most cases we are imagining passing functions from this package into others, so the multiple layers would get unduly complicated.

All that to say, I think we've got a solid idea of the issues and some potential approaches, and I look forward to figuring this out in person with you all at the conference.

**Comment by lmullen** [comment]
@kbenoit My understanding is that quanteda has its own reliable tokenizers now and that this issue can be closed. If that's not correct, please let me know and I'll come back to this.

**Comment by kbenoit** [comment]
**quanteda** does but we still have aspirations to outsource or modularize this. Happy for this to be closed though, since it's an open issue here https://github.com/quanteda/quanteda/issues/276.

**Comment by lmullen** [comment]
Okay, sounds like a plan. Happy to assist in making tokenizers suitable for quanteda.
